networkit/cpp/embedding/BiasedRandomWalk.cpp

```diff
@@ -0,0 +1,172 @@
+        graph.forNeighborsOf(v, [&](node x, edgeweight weight) {
+            if (x == t) {
+                pTable.push_back(weight / paramP);
+                pSum += weight / paramP;
+            } else if (tNbrs.count(x)) {
+                pTable.push_back(weight);
+                pSum += weight;
+            } else {
+                pTable.push_back(weight / paramQ);
+                pSum += weight / paramQ;
+            }
+        });
+        // Normalizing table
+        float pfSum = (float)pSum;
+        std::for_each(pTable.begin(), pTable.end(), [pfSum](float &p) { p /= pfSum; });
+
+        graphData->data[v][t].unigram(pTable);
+    }
+}
+
+std::vector<std::vector<node>> index2node;
+
+// Preprocess transition probabilities for each path t->v->x
+void preprocessTransitionProbs(const Graph &graph, double paramP, double paramQ) {
+    auto nn = graph.numberOfNodes();
+    graphData = std::make_unique<GraphData>(nn);
+
+    index2node.resize(nn);
+
+    // pre-allocate unordered maps:
+    for (index v = 0; v < nn; ++v) {
+        auto degv = graph.degreeOut(v);
+        for (auto n : graph.neighborRange(v)) {
+            // init index2node:
+            index2node[v].push_back(n);
+            graphData->data[v][n] = AliasSampler(degv);
+        }
+    }
+
+    graph.balancedParallelForNodes([&](node t) { preprocessNode(graph, t, paramP, paramQ); });
+}
+
+// Simulates a random walk
+Walk oneWalk(const Graph &graph, node start, count walkLen) {
+    Walk walk(walkLen);
+    count nr = 0;
+    walk[nr++] = start;
+    node src = start;
+
+    if (walkLen == 1) {
+        return walk;
+    }
+    if (graph.degreeOut(start) == 0) {
+        walk.resize(1); // shorten walk to 1
+        return walk;
+    }
+    auto nn = graph.degree(start);
+
+    node randomNeighbor = index2node[start][nn * uniform_real()];
+
+    walk[nr++] = randomNeighbor;
+    node dst = randomNeighbor;
+
+    while (nr < walkLen) {
+        if (graph.degreeOut(dst) == 0) {
+            walk.resize(nr); // shorten walk to nr
+            return walk;
+        }
+        NeighborMap &map = graphData->data[dst];
+        AliasSampler &as = map[src];
+        node next = index2node[dst][as.sample()];
+        walk[nr++] = next;
+        src = dst;
+        dst = next;
+    }
+    return walk;
+}
+
+struct WalkData {
+    // assuming contiguous node numbers 0..n-1, they serve as indices
+    AllWalks data;
+    count walkLength;
+    count walksPerNode;
+
+    WalkData(count nn, count wl, count wpn)
+        : data{nn * wpn, Walk(wl)}, walkLength(wl), walksPerNode(wpn) {}
+};
+
+/// Simulates walks from every node and writes it into Walks vector
+AllWalks doWalks(const Graph &graph, count walkLen, count walksPerNode) {
+    auto nn = graph.numberOfNodes();
+    count walksDone = 0;
+
+    WalkData walkData(nn, walkLen, walksPerNode);
+
+    std::vector<node> shuffled(nn);
+    node n = 0;
+    std::generate(shuffled.begin(), shuffled.end(), [&n]() { return n++; });
```

Comment:
To generate a vector with all the nodes of a graph it's better to use `std::vector<node> shuffled(graph.nodeRange().begin(), graph.nodeRange().end());`.